File: feature-timesheet/src/main/java/fr/o80/sample/timesheet/presentation/model/EntryViewModel.java
package fr.o80.sample.timesheet.presentation.model;

import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;

import java.util.List;

import fr.o80.sample.timesheet.data.entity.TimeEntry;

/**
 * @author Olivier Perez
 */
@AutoValue
public abstract class EntryViewModel {

    public static EntryViewModel inProgress() {
        return new AutoValue_EntryViewModel(true, false, null, null);
    }

    public static EntryViewModel success(List<TimeEntry> entries) {
        return new AutoValue_EntryViewModel(false, true, null, entries);
    }

    public static EntryViewModel error(Throwable throwable) {
        return new AutoValue_EntryViewModel(false, false, throwable, null);
    }

    public abstract boolean loading();

    public abstract boolean succeeded();

    @Nullable
    public abstract Throwable throwable();

    @Nullable
    public abstract List<TimeEntry> entries();
}

File: feature-timesheet/src/main/java/fr/o80/sample/timesheet/usecase/model/EntryViewModel.kt
package fr.o80.sample.timesheet.usecase.model

import android.os.Parcel
import android.os.Parcelable
import java.util.Date

/**
 * @author Olivier Perez
 */
data class EntryViewModel(val label: String, val code: String, val hours: Int, val date: Date) : Parcelable {
    constructor(parcel: Parcel) : this(
            parcel.readString(),
            parcel.readString(),
            parcel.readInt(),
            Date(parcel.readLong()))

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(label)
        parcel.writeString(code)
        parcel.writeInt(hours)
        parcel.writeLong(date.time)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<EntryViewModel> {
        override fun createFromParcel(parcel: Parcel): EntryViewModel {
            return EntryViewModel(parcel)
        }

        override fun newArray(size: Int): Array<EntryViewModel?> {
            return arrayOfNulls(size)
        }
    }
}
